File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # [0,1,2,3]
        l = None
        c = head
        while c:
            r = c.next
            c.next = l
            l = c
            c = r
        return l
        

File: Data Structures & Algorithms/reorder-linked-list/submission-0.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func reorderList(head *ListNode) {
	count := 0
	countP := head
	for countP != nil {
		count++
		countP = countP.Next
	}
	fmt.Println(count)
	mid := count / 2
	if count % 2 != 0 {
		mid++
	}
	fmt.Println(mid)
	// move mid ptr to mid-1 (need to cut 1st tail off)
	midP := head
	c := 0
	for c < mid-1 {
		midP = midP.Next
		c++
	}
	tmp := midP.Next
	midP.Next = nil
	midP = tmp
	fmt.Println(midP)
	// reverse the list
	fmt.Println("reversing")
	var prev *ListNode
	for midP != nil {
		temp := midP.Next
		midP.Next = prev
		prev = midP
		midP = temp
	}	

	fmt.Println("merging")
	first, second := head, prev
	for second != nil {
		temp1, temp2 := first.Next, second.Next
		first.Next = second
		second.Next = temp1
		first, second = temp1, temp2
	} 
}
